File: app/src/main/java/uk/co/sundroid/util/theme/ThemePalette.java
package uk.co.sundroid.util.theme;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Color;

import uk.co.sundroid.R.drawable;
import uk.co.sundroid.R.style;
import uk.co.sundroid.util.SharedPrefsHelper;
import uk.co.sundroid.util.astro.Body;
import uk.co.sundroid.util.astro.image.TrackerImage.TrackerStyle;

public class ThemePalette {

    public static final String THEME_DARK = "DARK";
    public static final String THEME_LIGHT = "LIGHT";
    private static String THEME = null;

    private static final TrackerStyle RADAR_LIGHT = new TrackerStyle(
            Color.argb(255, 0, 0, 0),
            Color.argb(100, 0, 0, 0),
            Color.argb(255, 255, 204, 0),
            Color.argb(255, 255, 168, 0),
            Color.argb(255, 72, 90, 144),
            Color.argb(255, 72, 90, 144),
            Color.argb(255, 255, 204, 0),
            Color.argb(255, 72, 90, 144),
            2,
            1,
            new float[] { 2, 3 },
            true
    );

    private static final TrackerStyle RADAR_DARK = new TrackerStyle(
            Color.argb(255, 255, 255, 255),
            Color.argb(170, 120, 120, 120),
            Color.argb(255, 255, 222, 107),
            Color.argb(255, 255, 198, 0),
            Color.argb(255, 72, 90, 144),
            Color.argb(255, 115, 142, 204),
            Color.argb(255, 255, 255, 255),
            Color.argb(255, 100, 100, 100),
            2,
            1,
            new float[] { 2, 3 },
            true
    );

    public static int getRiseArrow() {
        return THEME.equals(THEME_LIGHT) ? drawable.l_rise : drawable.d_rise;
    }

    public static int getSetArrow() {
        return THEME.equals(THEME_LIGHT) ? drawable.l_set : drawable.d_set;
    }

    public static int getAppBg() {
        return THEME.equals(THEME_LIGHT) ? drawable.l_app_bg : drawable.d_app_bg;
    }

    public static int getActionBarBg() {
        return THEME.equals(THEME_LIGHT) ? drawable.l_action_bar_bg : drawable.d_action_bar_bg;
    }

    public static int getPhaseFull() {
        return THEME.equals(THEME_LIGHT) ? drawable.l_phase_full : drawable.d_phase_full;
    }

    public static int getPhaseNew() {
        return THEME.equals(THEME_LIGHT) ? drawable.l_phase_new : drawable.d_phase_new;
    }

    public static int getPhaseLeft() {
        return THEME.equals(THEME_LIGHT) ? drawable.l_phase_left : drawable.d_phase_left;
    }

    public static int getPhaseRight() {
        return THEME.equals(THEME_LIGHT) ? drawable.l_phase_right : drawable.d_phase_right;
    }

    public static int getCalendarHighlightColor() {
        return THEME.equals(THEME_LIGHT) ? 0xffa6dff5 : 0xff222522;
    }

    public static int getCalendarDefaultColor() {
        return THEME.equals(THEME_LIGHT) ? 0x00ffffff : 0x00000000;
    }

    public static int getCalendarHeaderColor() {
        return THEME.equals(THEME_LIGHT) ? 0xffc4eaf8 : 0xff222522;
    }

    public static int getDisclosureOpen() {
        return THEME.equals(THEME_LIGHT) ? drawable.l_disclosure_open : drawable.d_disclosure_open;
    }

    public static int getDisclosureClosed() {
        return THEME.equals(THEME_LIGHT) ? drawable.l_disclosure_closed : drawable.d_disclosure_closed;
    }

    public static int getBodyColor(Body body) {
        return THEME.equals(THEME_LIGHT) ? body.getLightColor() : body.getDarkColor();
    }

    public static TrackerStyle getTrackerRadarStyle() {
        if (THEME.equals(THEME_LIGHT)) {
            return RADAR_LIGHT;
        } else {
            return RADAR_DARK;
        }
    }

    public static void changeToTheme(Activity activity, String theme) {
        THEME = theme;
        activity.finish();
        activity.startActivity(new Intent(activity, activity.getClass()));
    }

    public static void onActivityCreateSetTheme(Activity activity) {
        if (THEME == null) {
            String prefsTheme = SharedPrefsHelper.getTheme(activity);
            if (prefsTheme.equals(THEME_DARK) || prefsTheme.equals(THEME_LIGHT)) {
                THEME = prefsTheme;
            }
        }
        if (THEME.equals(THEME_LIGHT)) {
            activity.setTheme(style.SundroidLight);
        } else {
            activity.setTheme(style.SundroidDark);
        }
    }

}

File: app/src/main/kotlin/uk/co/sundroid/util/theme/ThemePalette.kt
package uk.co.sundroid.util.theme

import android.app.Activity
import android.graphics.Color
import uk.co.sundroid.R

import uk.co.sundroid.R.style
import uk.co.sundroid.util.prefs.Prefs
import uk.co.sundroid.util.astro.Body
import uk.co.sundroid.util.astro.image.TrackerImage.TrackerStyle
import uk.co.sundroid.R.drawable.*

const val THEME_DARK = "DARK"
const val THEME_DARKBLUE = "DARKBLUE"
var theme: String? = null

private val radarDark = TrackerStyle(
        Color.argb(255, 255, 255, 255),
        Color.argb(170, 120, 120, 120),
        Color.argb(255, 255, 204, 0),
        Color.argb(255, 255, 157, 0),
        Color.argb(255, 99, 116, 166),
        Color.argb(255, 72, 90, 144),
        Color.argb(255, 47, 65, 119),
        Color.argb(255, 26, 41, 88),
        Color.argb(255, 255, 255, 255),
        Color.argb(255, 92, 118, 168),
        Color.argb(255, 255, 255, 255),
        Color.argb(0, 0, 0, 0),
        3,
        2,
        true
)

fun appBackground(): Int {
    return when (theme) {
        THEME_DARK -> R.color.d_app_background
        else -> R.color.db_app_background
    }
}

fun getRiseArrow(): Int {
    return d_rise
}

fun getSetArrow(): Int {
    return when (theme) {
        THEME_DARKBLUE -> db_set
        else -> d_set
    }
}

fun getRisenAllDay(): Int {
    return d_risen_all_day
}

fun getSetAllDay(): Int {
    return when (theme) {
        THEME_DARKBLUE -> db_set_all_day
        else -> d_set_all_day
    }
}

//fun getAppBg(): Int {
//    return when (theme) {
//        THEME_DARKBLUE -> db_app_bg
//        else -> d_app_bg
//    }
//}

fun getPhaseFull(): Int {
    return d_phase_full
}

fun getPhaseNew(): Int {
    return d_phase_new
}

fun getPhaseLeft(): Int {
    return d_phase_left
}

fun getPhaseRight(): Int {
    return d_phase_right
}

fun getCalendarGridHighlightColor(): Int {
    return when (theme) {
        THEME_DARKBLUE -> 0xff162544.toInt()
        else -> 0xff323632.toInt()
    }
}

fun getCalendarGridDefaultColor(): Int {
    return when (theme) {
        THEME_DARKBLUE -> 0x810D1629.toInt()
        else -> 0x81222522.toInt()
    }
}

fun getCalendarHeaderColor(): Int {
    return when (theme) {
        THEME_DARKBLUE -> 0xff0D1629.toInt()
        else -> 0xff222522.toInt()
    }
}

fun upColor(): Int {
    return 0xffffffff.toInt()
}

fun downColor(): Int {
    return when (theme) {
        THEME_DARKBLUE -> 0xff5C76A8.toInt()
        else -> 0xff8a918a.toInt()
    }
}

fun getDisclosureOpen(): Int {
    return d_disclosure_open
}

fun getDisclosureClosed(): Int {
    return d_disclosure_closed
}

fun getBodyColor(body: Body): Int {
    return body.darkColor
}

fun getTrackerRadarStyle(): TrackerStyle {
    return radarDark
}

fun changeToTheme(activity: Activity, newTheme: String) {
    theme = newTheme
    activity.recreate()
}

fun onActivityCreateSetTheme(activity: Activity) {
    if (theme == null) {
        val prefsTheme = Prefs.theme(activity)
        if (prefsTheme == THEME_DARK || prefsTheme == THEME_DARKBLUE) {
            theme = prefsTheme
        }
    }
    when (theme) {
        THEME_DARKBLUE -> activity.setTheme(style.SundroidDarkBlue)
        else -> activity.setTheme(style.SundroidDark)
    }
}
